python/services/dataservice/dmod/dataservice/data_derive_util.py

```diff
@@ -0,0 +1,317 @@
+
+        Returns
+        -------
+        bool
+            Whether deriving an appropriate realization configuration is possible.
+        """
+        if requirement.category != DataCategory.CONFIG:
+            return False
+        elif requirement.domain.data_format != DataFormat.NGEN_REALIZATION_CONFIG:
+            return False
+
+        request = job.model_request
+        return (isinstance(request, AbstractNgenRequest) and request.is_intelligent_request
+                and isinstance(request.formulation_configs, PartialRealizationConfig))
+
+    async def derive_datasets(self, job: Job) -> List[DataRequirement]:
+        """
+        Derive any datasets as required for the given job awaiting its data.
+
+        Job is expected to be in the ``AWAITING_DATA`` status step.  If it is not, no datasets are derived, a warning is
+        logged, and an empty list is returned.
+
+        If in the right status, but initially any unfulfilled requirements of the job cannot have a dataset successfully
+        derived, a ::class:`DmodRuntimeError` is raised.
+
+        Parameters
+        ----------
+        job : Job
+            The job having ::class:`DataRequirement` objects for which fulfilling datasets should be derived.
+
+        Returns
+        -------
+        List[DataRequirement]
+            A list of the given job's data requirements for a which a fulfilling dataset was derived and associated.
+
+        Raises
+        -------
+        DmodRuntimeError
+            Raised if a job with the correct status has an initially unfulfilled requirement for which a satisfactory
+            dataset can not be derived by this function.
+        """
+        # Only do something if the job has the right status
+        if job.status_step != JobExecStep.AWAITING_DATA:
+            return []
+
+        results = []
+
+        # TODO: #needs_issue Make sure logic is in place to make derived datasets temporary (unless somehow specified to
+        #  not be temporary) and have temporary datasets cleaned up by service periodically
+        # TODO: #needs_issue Also make sure that temporary datasets have their expire time updated if they are used
+        #  again for something
+        for req in [r for r in job.data_requirements if r.fulfilled_by is None]:
+            # **********************************************************************************************************
+            # *** TODO: if/when deriving forcing datasets is supported, make sure this is done before config datasets
+            # *** TODO: when generating BMI datasets is supported, make sure it's done before realization configs
+            # **********************************************************************************************************
+            # Derive realization config datasets from formulations in message body when necessary
+            if req.category == DataCategory.CONFIG and req.domain.data_format == DataFormat.NGEN_REALIZATION_CONFIG:
```

Comment:
I am not following exactly what you are suggesting as an alternative.  What specific rollback scenarios/support did you have in mind?  

I don't initially see a problem with, say, allowing a job requirement to keep the fulfillment (and location) association we've just set with a newly-derived dataset, even if a subsequent requirement fails to derive something.  The `_derive_realization_config_from_formulations` function only makes those assignments after the data is added to a new dataset successfully (I expect later, similar functions to work the same way).  In other words, those datasets should exist, have satisfactory domains, and contain data consistent with their domains, so the dataset should fulfill the requirement just fine.

We could discuss whether `_derive_realization_config_from_formulations` should have side effects or return values.  That would impact where and how we could implement rollback capabilities, but wouldn't imply any rollback functionality changes directly.

We probably should roll back on the creation of a dataset within the `_derive_realization_config_from_formulations` function  if adding data to it fails.  I'll work on a change for that shortly.  

It occurs to me that if we treat things as temporary, and can retry/restart jobs,  we could have a scenario with a temp dataset being deleted, although I would expect that to take a lot of time going by to actually happen (i.e., something rare enough that we need to deal with it eventually, but not now)

Reply:
> I am not following exactly what you are suggesting as an alternative. What specific rollback scenarios/support did you have in mind?

I shouldnt have used the term _rollback_, it doesnt do the _best_ job at characterizing what I was trying to convey. Sorry for my poor word choice. The root of what I was trying to communicate is that derived dataset creation should not be dependent on job data requirement ordering. Meaning, it would seem that we would _not_ want to create any derived datasets if an unsupported data requirement exists in the list of job data requirements. This could lead to situations where it is unclear why a dataset was created and if creation of that dataset was intended or just user error.